### PR TITLE
feat(openapi): implement getting binary file from OSS

### PR DESCRIPTION
**Pull request by graczhual**

### CODE REVIEW

**Comment by GravitiRob** [comment]
Task linked: [CU-2meww9b 二进制文件读取对接: OSS](https://app.clickup.com/t/2meww9b)

File: graviti/openapi/policy.py [hunk 1]
```diff
@@ -92,3 +184,58 @@ def get_object_policy(
...
+                policy["AccessKeyId"],
+                policy["AccessKeySecret"],
+                policy["SecurityToken"],
+                policy["endpoint"],
+                policy["bucket"],
+                file_name,
+            )
+        except ResponseError as error:
+            _error = xmltodict.parse(error.response.text)["Error"]
+            if allow_retry and _error["Code"] == "AccessDenied":
```
Comment (Lee-000) [inline graviti/openapi/policy.py:237]: Could the file be retried after retry with the same params when got `AccessDenied` error?

File: graviti/openapi/policy.py [hunk 2]
```diff
@@ -5,9 +5,101 @@
...
+    permission["expireAt"] = convert_iso_to_datetime(permission["policy"].pop("expireAt"))
+    _PERMISSIONS[access_key, owner, dataset, "GET"] = permission
+
+
+def _get_read_permission(
+    access_key: str,
+    url: str,
+    owner: str,
+    dataset: str,
+    request_again: bool,
```
Comment (Lee-000) [inline graviti/openapi/policy.py:50]: Is there a better name?

File: graviti/openapi/policy.py [hunk 3]
```diff
@@ -5,9 +5,101 @@
...
+
+
+@locked
+def _request_read_permission(
+    access_key: str,
+    url: str,
+    owner: str,
+    dataset: str,
+) -> None:
+    permission: Dict[str, Any] = get_object_policy(
```
Comment (Lee-000) [inline graviti/openapi/policy.py:33]: `permission` or `policy`?

Comment (graczhual) [reply graviti/openapi/policy.py:33]: the response of get_object_policy is the dict like `{"backend_type": "OSS", "policy": {...}}`, so we could not call this response as `policy`

File: graviti/openapi/policy.py (hunk 2 above)
Comment (graczhual) [reply graviti/openapi/policy.py:50]: `need_request_again`?

File: graviti/openapi/policy.py [hunk 4]
```diff
@@ -5,9 +5,101 @@
...
+
+
+def _get_binary_file_from_oss(  # pylint: disable=too-many-arguments
+    access_key_id: str,
+    access_key_secret: str,
+    security_token: str,
+    endpoint: str,
+    bucket: str,
+    object_name: str,
+) -> ByteString:
```
Comment (linjiX) [inline graviti/openapi/policy.py:91]: why the return type is ByteString, can we get a FilePointer like object from OSS and use read to lazy load the binary content?

File: graviti/openapi/policy.py [hunk 5]
```diff
@@ -5,9 +5,101 @@
...
+    access_key: str,
+    url: str,
+    owner: str,
+    dataset: str,
+    request_again: bool,
+) -> Dict[str, Any]:
+    key = (access_key, owner, dataset, "GET")
+    if (
+        request_again
+        or datetime.now(timezone.utc) >= _PERMISSIONS.get(key, _DEFAULT_PERMISSION)["expireAt"]
```
Comment (linjiX) [inline graviti/openapi/policy.py:55]: will it be better to send request first, use the error message to check the expire issue, then ask the permission and rerequest

File: graviti/openapi/policy.py [hunk 6]
```diff
@@ -92,3 +184,58 @@ def get_object_policy(
...
+            return _get_binary_file_from_oss(
+                policy["AccessKeyId"],
+                policy["AccessKeySecret"],
+                policy["SecurityToken"],
+                policy["endpoint"],
+                policy["bucket"],
+                file_name,
+            )
+        except ResponseError as error:
+            _error = xmltodict.parse(error.response.text)["Error"]
```
Comment (linjiX) [inline graviti/openapi/policy.py:236]: better to not use third party to parse xml, python builtin xml library is an option.
Also can we ask OSS to return error message in other format like JSON.

File: graviti/openapi/policy.py [hunk 7]
```diff
@@ -5,9 +5,101 @@
...
+from graviti.utility import config, convert_iso_to_datetime, locked
+
+_EXPIRED_IN_SECOND = 240
+_GMT_FORMAT = "%a, %d %b %Y %H:%M:%S GMT"
+_PERMISSIONS: Dict[Tuple[str, str, str, str], Dict[str, Any]] = {}
+_DEFAULT_PERMISSION = {"expireAt": datetime.fromtimestamp(0, timezone.utc)}
+
+
+@locked
+def _request_read_permission(
```
Comment (linjiX) [inline graviti/openapi/policy.py:27]: will it be better to use a class rather than functions.

File: graviti/openapi/policy.py [hunk 8]
```diff
@@ -5,9 +5,133 @@
...
+
+class ObjectPolicy:
+    """The basic structure of the object policy of the dataset.
+
+    Arguments:
+        dataset: Class :class:`~graviti.dataset.dataset.Dataset` instance.
+
+    """
+
+    _read_policy: Optional[Dict[str, Any]] = None
```
Comment (linjiX) [inline graviti/openapi/policy.py:35]: will `_get_policy` better?

File: graviti/openapi/policy.py [hunk 9]
```diff
@@ -5,9 +5,133 @@
...
+            dataset.url,
+            dataset.owner,
+            dataset.name,
+            actions="GET",
+            is_internal=config.is_internal,
+            expired=_EXPIRED_IN_SECOND,
+        )["policy"]
+
+    @property
+    def read_policy(self) -> Dict[str, Any]:
```
Comment (linjiX) [inline graviti/openapi/policy.py:54]: use method rather than property

File: graviti/openapi/policy.py [hunk 10]
```diff
@@ -5,9 +5,133 @@
...
+        """Return the read policy.
+
+        Returns:
+            The read policy.
+
+        """
+        if self._read_policy is None:
+            self._request_read_policy()
```
Comment (linjiX) [inline graviti/openapi/policy.py:62]: `self._read_policy = self._request_read_policy()`?

File: graviti/openapi/policy.py [hunk 11]
```diff
@@ -5,9 +5,133 @@
...
+        return self._read_policy  # type: ignore[return-value]
+
+
+class OssObjectPolicy(ObjectPolicy):
+    """The basic structure of the object policy of the dataset stored in oss."""
+
+    _read_hashing: hmac.HMAC
```
Comment (linjiX) [inline graviti/openapi/policy.py:69]: `_read_hmac`?

File: graviti/openapi/policy.py [hunk 12]
```diff
@@ -5,9 +5,133 @@
...
+
+    def _request_read_policy(self) -> None:
+        super()._request_read_policy()
+        self._read_hashing = hmac.new(
+            self._read_policy["AccessKeySecret"].encode("utf-8"),  # type: ignore[index]
+            digestmod=hashlib.sha1,
+        )
+
+    def _get_oss_authorization(
```
Comment (linjiX) [inline graviti/openapi/policy.py:78]: can we remove `oss` in the method name?

File: graviti/openapi/policy.py [hunk 13]
```diff
@@ -5,9 +5,133 @@
...
+        self,
+        policy: Dict[str, Any],
+        verb: str,
+        date: str,
+        object_name: str,
```
Comment (linjiX) [inline graviti/openapi/policy.py:83]: `key` or `object_name`?

File: graviti/openapi/policy.py [hunk 14]
```diff
@@ -5,9 +5,133 @@
...
+    ) -> str:
+        signature = (
+            f"{verb}\n\n\n{date}\nx-oss-security-token:{policy['SecurityToken']}\n"
+            f"/{policy['bucket']}/{object_name}"
+        )
+        hashing = self._read_hashing.copy()
```
Comment (linjiX) [inline graviti/openapi/policy.py:89]: hashing -> hmac?

File: graviti/openapi/policy.py [hunk 15]
```diff
@@ -5,9 +5,133 @@
...
+        url = f"https://{policy['bucket']}.{policy['endpoint']}/{object_name}"
+
+        try:
+            response = do(verb, url, headers=headers, timeout=config.timeout, stream=True)
+            return UserResponse(response)
+        except ResponseError as error:
+            code = ElementTree.fromstring(error.response.text)[0].text
+            if allow_retry and code in _OSS_RETRY_CODE:
+                delattr(self, "_read_policy")
+                self.read_object(object_name, False)
```
Comment (linjiX) [inline graviti/openapi/policy.py:131]: `return self.read_object(object_name, False)`

File: graviti/openapi/policy.py [hunk 16]
```diff
@@ -5,9 +5,133 @@
...
+        hashing = self._read_hashing.copy()
+        hashing.update(signature.encode("utf-8"))
+        signature = base64.b64encode(hashing.digest()).decode("utf-8")
+        return f"OSS {policy['AccessKeyId']}:{signature}"
+
+    def read_object(
+        self,
+        object_name: str,
+        allow_retry: bool = True,
+    ) -> Optional[UserResponse]:
```
Comment (linjiX) [inline graviti/openapi/policy.py:98]: the return type should not be `Optional`

File: graviti/openapi/policy.py [hunk 17]
```diff
@@ -5,9 +5,133 @@
...
+        }
+        url = f"https://{policy['bucket']}.{policy['endpoint']}/{object_name}"
+
+        try:
+            response = do(verb, url, headers=headers, timeout=config.timeout, stream=True)
+            return UserResponse(response)
+        except ResponseError as error:
+            code = ElementTree.fromstring(error.response.text)[0].text
+            if allow_retry and code in _OSS_RETRY_CODE:
+                delattr(self, "_read_policy")
```
Comment (linjiX) [inline graviti/openapi/policy.py:130]: will it be better to add a method named `clear_read_policy`?

File: graviti/openapi/policy.py [hunk 18]
```diff
@@ -5,9 +5,133 @@
...
+        Returns:
+            The read policy.
+
+        """
+        if self._read_policy is None:
+            self._request_read_policy()
+        return self._read_policy  # type: ignore[return-value]
+
+
+class OssObjectPolicy(ObjectPolicy):
```
Comment (linjiX) [inline graviti/openapi/policy.py:66]: should the `Oss` be all uppercase?

**Review by linjiX — CHANGES_REQUESTED** [review]

**Review by linjiX — APPROVED** [review]

**Review by Lee-000 — APPROVED** [review]

**Pull request merged**